plugins/trace_api_plugin/compressed_file.cpp

```diff
@@ -0,0 +1,311 @@
+            return;
+         }
+      }
+
+
+      // decompress more chunks
+      while (written < n) {
+         if ( strm.avail_in == 0 ) {
+            size_t remaining = file_size - file.tellp();
+            size_t to_read = std::min((size_t)compressed_buffer.size(), remaining);
+            file.read(reinterpret_cast<char*>(compressed_buffer.data()), to_read);
+            strm.avail_in = to_read;
+            strm.next_in = compressed_buffer.data();
+         }
+
+         do {
+            strm.avail_out = read_buffer.size();
+            strm.next_out = read_buffer.data();
+            auto ret = inflate(&strm, Z_NO_FLUSH);
+
+            if (ret == Z_NEED_DICT || ret == Z_DATA_ERROR || ret == Z_MEM_ERROR) {
+               throw compressed_file_error("Error decompressing: " + std::string(strm.msg));
+            }
+
+            auto bytes_decompressed = read_buffer.size() - strm.avail_out;
+            auto to_copy = std::min(bytes_decompressed, n - written);
+            std::memcpy(d + written, read_buffer.data(), to_copy);
+            written += to_copy;
+
+            if (bytes_decompressed > to_copy) {
+               // move remaining to the front of the buffer
+               std::memmove(read_buffer.data(), read_buffer.data() + to_copy, bytes_decompressed - to_copy);
+               remaining_read_buffer = bytes_decompressed - to_copy;
+            }
+
+            if (written < n && ret == Z_STREAM_END) {
+               throw std::ios_base::failure("Attempting to read past the end of a compressed file");
+            }
+         } while (strm.avail_out == 0 && written < n);
+      }
+   }
+
+   void seek( long loc, fc::cfile& file ) {
+      if (initialized) {
+         inflateEnd(&strm);
+         initialized = false;
+      }
+
+      // read in the seek point map
+      file.seek_end(-2);
+      uint16_t seek_point_count = 0;
+      file.read(reinterpret_cast<char*>(&seek_point_count), 2);
+
+      int seek_map_size = 16 * seek_point_count;
+      file.seek_end(-2 - seek_map_size);
+
+      std::vector<seek_point_entry> seek_point_map(seek_point_count);
+      file.read(reinterpret_cast<char*>(seek_point_map.data()), seek_point_map.size() * sizeof(seek_point_entry));
+
+      // seek to the neareast seek point
+      auto iter = std::lower_bound(seek_point_map.begin(), seek_point_map.end(), (uint64_t)loc, []( const auto& lhs, const auto& rhs ){
+         return std::get<0>(lhs) < rhs;
+      });
+
+      // special case when there is a seek point that is exact
+      if ( iter != seek_point_map.end() && std::get<0>(*iter) == loc ) {
+         file.seek(std::get<1>(*iter));
+         return;
+      }
+
+      long remaining = loc;
+
+      // special case when this is before the first seek point
+      if ( iter == seek_point_map.begin() ) {
+         file.seek(0);
+      } else {
+         // if lower bound wasn't exact it will be one past the seek point we need
```

Comment:
got confused on this, I think changing it to "iter" instead of "it" will make it clearer.

Reply:
What I wasn't sure about here is if it is safe not to check for `seek_point_map.end()`. I think I convinced myself that is not possible but I wouldn't stake my life on it.